### PR TITLE
Internal SPA Links

### LINKED ISSUE #542: Integrate react router links for internal links
**Is your feature request related to a problem? Please describe.**
We currently do not make use of our SPA because all of our links are normal hyperlinks.

**Describe the solution you'd like**
Use React Router links based on regex that matches our internal routes, i.e. `/news/articles/c0000000025o`.

### PR DESCRIPTION
**Pull request by jtart**
Resolves #542 

_Uses a react-router link if the link URL matches regex, otherwise uses a normal hyperlink._

_When the client-side React application takes control, there needs to be some handling of async data fetching. After.JS does not give states in the async data fetching lifecycle. This can be mitigated against by checking for a lack of data, which is fairly crude but there are limited options available. A lack of data can be, and is in the PR, used for a loading state, but will **NOT** handle errors, which is a huge limitation of After.JS._

- [x] Tests added for new features
- [ ] Test engineer approval

### CODE REVIEW

File: src/app/containers/InlineLink/index.jsx [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import React from 'react';
+import { Link } from 'react-router-dom';
+import { string, node } from 'prop-types';
+import InlineLink from '../../components/InlineLink';
+
+const InternalInlineLink = InlineLink.withComponent(Link);
+
+const InlineLinkContainer = ({ href, children, ...rest }) => {
+  // if URL matches a valid route, use a react-router link
+  if (href.match('^/news/articles/c[a-zA-Z0-9]{10}o$')) {
```
Comment (sareh) [inline src/app/containers/InlineLink/index.jsx]: You should move this regex out to a constant in a library file, since we use it in this file too `/containers/Article/index.jsx` and will likely need to re-use it in other places as well.

Comment (sareh) [reply src/app/containers/InlineLink/index.jsx]: Also this check `is a valid article route` should be a library function that we import and use in this file.

File: src/app/containers/InlineLink/index.test.jsx [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import React from 'react';
+import { StaticRouter } from 'react-router-dom';
+import { shouldMatchSnapshot } from '../../helpers/tests/testHelpers';
+import InlineLink from './index';
+
+describe('InlineLink', () => {
+  describe('internal link', () => {
```
Comment (sareh) [inline src/app/containers/InlineLink/index.test.jsx]: Renaming these descriptions would add clarity. 
In most cases we use "internal link" to mean "a link going to another BBC page" but here you're referring to "a link within this SPA". Same for "external link" below.

File: src/app/containers/Text/index.test.jsx [hunk 3]
```diff
@@ -35,6 +36,11 @@ describe('TextContainer', () => {
       ],
     };
 
-    shouldMatchSnapshot('should render correctly', <TextContainer {...data} />);
+    shouldMatchSnapshot(
+      'should render correctly',
+      <StaticRouter>
```
Comment (sareh) [inline src/app/containers/Text/index.test.jsx]: An explanatory comment about why `StaticRouter` is needed here would be useful, for devs who maintain these tests in future.

File: src/app/containers/Article/index.jsx [hunk 4]
```diff
@@ -35,6 +35,16 @@ class ArticleContainer extends Component {
...
   render() {
     const { data } = this.props;
+
+    /*
+      This handles async data fetching, and a 'loading state', which we should look to handle more intelligently.
+      After.JS gives no explicit loading state, so we just have to check for a lack of data.
+      After.JS should handle async data fetching in a more informative way, which will be become an issue for error handling.
+    */
+    if (!data) {
+      return 'Loading...';
```
Comment (sareh) [inline src/app/containers/Article/index.jsx:45]: This should be a separate PR, alongside the change here https://github.com/bbc/simorgh/pull/538/files#diff-b83d3a0f6ef43a3dcaea2ca5421e2743L48

File: src/app/components/Text/index.jsx [hunk 5]
```diff
@@ -3,7 +3,7 @@ import styled from 'styled-components';
 import { string, bool, shape, node } from 'prop-types';
 import Markdown from 'markdown-to-jsx';
 import mediaQuery from '../../helpers/mediaQueries';
-import InlineLink from '../InlineLink';
+import InlineLink from '../../containers/InlineLink';
```
Comment (sareh) [inline src/app/components/Text/index.jsx:6]: This is more of a question about our application architecture - do we want components calling containers? 
I'd have thought the `Text` container should import the `InlineLink` container and pass it to the `Text` component.

**Review by sareh — CHANGES_REQUESTED** [review]
Great work. I've just got a few comments and questions.

File: src/app/containers/Article/index.jsx (hunk 4 above)
Comment (jtart) [reply src/app/containers/Article/index.jsx:45]: It could very well be in a PR, but that PR would have to be merged before implementing React Router Links. I'm happy to do that work, but we should also consider whether that work would be valuable when we are looking to get this work through fairly quickly.

File: src/app/components/Text/index.jsx (hunk 5 above)
Comment (jtart) [reply src/app/components/Text/index.jsx:6]: I see your point, and might come into a wider question on component design architecture that is out of scope of this PR. We already do something similar to this in other components.

File: src/app/containers/InlineLink/index.jsx (hunk 1 above)
Comment (jtart) [reply src/app/containers/InlineLink/index.jsx]: Would you accept doing all this in a follow-up PR?

File: src/app/components/Text/index.jsx (hunk 5 above)
Comment (sareh) [reply src/app/components/Text/index.jsx:6]: @jtart I have opened an issue to address this https://github.com/bbc/simorgh/issues/543

**Review by sareh — DISMISSED** [review]
👍 Looks good.

File: src/app/components/Text/index.jsx (hunk 5 above)
Comment (jp-bbc) [reply src/app/components/Text/index.jsx:6]: @sareh I would agree with your point @sareh. However, wouldn't this approach be considered an anti-pattern since they separate two differing concerns?

File: src/app/containers/Article/index.jsx (hunk 4 above)
Comment (jp-bbc) [reply src/app/containers/Article/index.jsx:45]: This feels a fairly large PR. I would break it down into a separate PR.

Comment (jtart) [reply src/app/containers/Article/index.jsx:45]: @jp-bbc can you give points where it should be split into smaller tasks? Please also see a justification for not doing that in the comment above.

**Comment by pjlee11** [comment]
Assuming the points mentioned throughout other reviews are going to be addressed in a follow up PR then happen with this 👍

**Review by pjlee11 — DISMISSED** [review]

**Review by pjlee11 — DISMISSED** [review]

**Review by sareh — DISMISSED** [review]

File: cypress/integration/bodySpec.js [hunk 6]
```diff
@@ -50,4 +51,8 @@ describe('Article Body Tests', () => {
   it('should render a title', () => {
     cy.title().should('eq', "Meghan's bouquet laid on tomb of unknown warrior");
   });
+
+  it('should have working inline links', () => {
+    clickInlineLinkAndTestPageHasHTML('main a', '/news/articles/c0000000027o');
```
Comment (sareh) [inline cypress/integration/bodySpec.js]: We should update the description here and of the helper utility to be descriptive.

**Review by sareh — APPROVED** [review]

**Review by pjlee11 — APPROVED** [review]
